web/packages/teleport/src/AppLauncher/AppLauncher.test.tsx

```diff
@@ -52,3 +50,75 @@ test('arn is url decoded', () => {
     arn: 'arn:aws:iam::joe123:role/EC2FullAccess',
   });
 });
+
+const testCases = [
+  {
+    queryParams: '?path=%2F',
+    expectedPath: '/',
+  },
+  {
+    queryParams: '?path=%2Flogin',
+    expectedPath: '/login',
+  },
+  {
+    queryParams: '?path=%2Ffoo%2Fbar&fruit=apple&os=mac',
+    expectedPath: '/foo/bar?fruit=apple&os=mac',
+  },
+  {
+    queryParams: '?path=',
+    expectedPath: '/',
+  },
+  {
+    queryParams: '?path=&fruit=apple',
+    expectedPath: '/?fruit=apple',
+  },
+  {
+    queryParams:
+      '?path=%2Falerting%2Flist&search=state:pending%20type:recording%20health:error',
+    expectedPath:
+      '/alerting/list?search=state:pending+type:recording+health:error',
+  },
+];
+
+describe('app launcher path is properly formed', () => {
+  const realLocation = window.location;
+  const assignMock = jest.fn();
+
+  beforeEach(() => {
+    global.fetch = jest.fn(() => Promise.resolve({})) as jest.Mock;
+    jest.spyOn(api, 'get').mockResolvedValue({});
+    jest.spyOn(api, 'post').mockResolvedValue({});
+
+    delete window.location;
+    window.location = { ...realLocation, replace: assignMock };
+  });
+
+  afterEach(() => {
+    window.location = realLocation;
+    assignMock.mockClear();
+  });
+
+  test.each(testCases)(
+    '$queryParams',
+    async ({ queryParams, expectedPath }) => {
+      const launcherPath = `/web/launch/grafana.localhost${queryParams}`;
+      const mockHistory = createMemoryHistory({
+        initialEntries: [launcherPath],
+      });
+
+      render(
+        <Router history={mockHistory}>
+          <Route path={cfg.routes.appLauncher}>
+            <AppLauncher />
+          </Route>
+        </Router>
+      );
+
+      await waitFor(() =>
+        expect(window.location.replace).toHaveBeenCalledWith(
+          `https://grafana.localhost${expectedPath}`
+        )
+      );
```

Comment:
test works, it passes when expected matches, and fails if it doesn't, but i can't figure out why i get this error. it happens because for some reason, the [callback](https://github.com/gravitational/teleport/blob/aea9f593089e6c68bfc18b7e28f3926b78937c00/web/packages/teleport/src/AppLauncher/AppLauncher.tsx#L37) is called twice (but not for each case), and the second call fails to make a network request? but when I manually test it, it only gets called once.

still trying to figure it out in the mean time

<img width="1408" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/0ae49af0-d79b-4cdf-8670-f60166624dcc">

Reply:
it was the existing test's error messages, i shifted it down into my describe block and no more console err msg's